system/jlib/jfcmp.hpp

```diff
@@ -0,0 +1,516 @@
+    size32_t outBufStart;
+    byte *inbuf;
+    size32_t inmax;         // remaining
+    size32_t inlen;
+    size32_t inlenblk;      // set to COMMITTED when so
+    bool trailing;
+    byte *outbuf;
+    size32_t outlen;
+    size32_t wrmax;
+    size32_t dynamicOutSz;
+
+    virtual inline void setinmax()
+    {
+        inmax = blksz-outlen-2*sizeof(size32_t);
+        if (inmax<256)
+            trailing = true;    // too small to bother compressing
+        else
+            trailing = false;
+    }
+
+    virtual inline void flushcommitted()
+    {
+        // only does non trailing
+        if (trailing)
+            return;
+        size32_t toflush = (inlenblk==COMMITTED)?inlen:inlenblk;
+        if (toflush == 0)
+            return;
+
+        if (toflush < 256)
+        {
+            trailing = true;
+            return;
+        }
+
+        size32_t outSzRequired = outlen+sizeof(size32_t)*2+toflush;
+        if (!dynamicOutSz)
+            assertex(outSzRequired<=blksz);
+        else
+        {
+            if (outSzRequired>dynamicOutSz)
+            {
+                verifyex(outBufMb->ensureCapacity(outBufStart+outSzRequired));
+                dynamicOutSz = outBufMb->capacity();
+                outbuf = ((byte *)outBufMb->bufferBase()+outBufStart);
+            }
+        }
+        size32_t *cmpsize = (size32_t *)(outbuf+outlen);
+        byte *out = (byte *)(cmpsize+1);
+
+        memcpy(out, inbuf, toflush);
+        *cmpsize = toflush;
+
+        *(size32_t *)outbuf += toflush;
+        outlen += *cmpsize+sizeof(size32_t);
+        if (inlenblk==COMMITTED)
+            inlen = 0;
+        else
+        {
+            inlen -= inlenblk;
+            memmove(inbuf,inbuf+toflush,inlen);
+        }
+        setinmax();
+        return;
+    }
+
+    void initCommon()
+    {
+        blksz = inma.capacity();
+        *(size32_t *)outbuf = 0;
+        outlen = sizeof(size32_t);
+        inlen = 0;
+        inlenblk = COMMITTED;
+        setinmax();
+    }
+
+public:
+    CFcmpCompressor()
+    {
+        outlen = 0;
+        outbuf = NULL;      // only set on close
+        bufalloc = 0;
+        wrmax = 0;          // set at open
+        dynamicOutSz = 0;
+        outBufMb = NULL;
+        outBufStart = 0;
+        inbuf = NULL;
+    }
+
+    virtual ~CFcmpCompressor()
+    {
+        if (bufalloc)
+            free(outbuf);
+    }
+
+
+    virtual void open(void *buf,size32_t max)
+    {
+        if (max<1024)
+            throw MakeStringException(-1,"CFcmpCompressor::open - block size (%d) not large enough", blksz);
+        wrmax = max;
+        if (buf)
+        {
+            if (bufalloc)
+                free(outbuf);
+            bufalloc = 0;
+            outbuf = (byte *)buf;
+        }
+        else if (max>bufalloc)
+        {
+            if (bufalloc)
+                free(outbuf);
+            bufalloc = max;
+            outbuf = (byte *)malloc(bufalloc);
```

Comment:
minor: should validate succeeded really and throw error if not and only set bufalloc=max afterwards